### PR TITLE
Add channel_id to AllToAllOp

### PR DESCRIPTION
**Pull request by burmako**
Rationale for approval is the same as in #272, #388, #403 and #673: this is a non-controversial backward-compatible change, and accepting it doesn't violate any of the existing commitments (it sticks to existing HLO semantics, and it is compatible with the extent of the current compatibility commitments).

MLIR-HLO commit: tensorflow/mlir-hlo@bd07cb9.

### CODE REVIEW

File: stablehlo/transforms/VhloToVersion.cpp [hunk 1]
```diff
@@ -326,6 +348,8 @@ void populateVhloToVersionPatterns(RewritePatternSet* patterns,
   patterns->add<vhlo::CollectivePermuteOpV2ToV1>(*converter, context);
   patterns->add<vhlo::AllGatherOpV1ToV2>(*converter, context);
   patterns->add<vhlo::AllGatherOpV2ToV1>(*converter, context);
+  patterns->add<vhlo::AllToAllOpV1ToV2>(*converter, context);
```
Comment (GleasonK) [inline stablehlo/transforms/VhloToVersion.cpp:351]: Doesn't have to be here, but I think we can add multiple patterns in a single invocation. Would help pair the upgrade/downgrade:

```
patterns->add<vhlo::AllToAllOpV1ToV2, vhlo::AllToAllOpV2ToV1>(*converter, context);
```

File: stablehlo/tests/vhlo_to_version_downgrade_invalid.mlir [hunk 2]
```diff
@@ -80,3 +80,18 @@ func.func @invalid_program_unknown_op(%arg0 : tensor<f32>) -> (tensor<f32>) {
...
+
+func.func @all_to_all_to_v1(%arg0: tensor<4x16xf32>) -> tensor<16x4xf32> {
+  // expected-error @+2 {{failed to downgrade vhlo.all_to_all_v2, op has a non-empty channel_handle attribute}}
+  // expected-error @+1 {{failed to legalize operation 'vhlo.all_to_all_v2' that was explicitly marked illegal}}
+  %0 = "stablehlo.all_to_all"(%arg0) {
+    split_dimension = 1 : i64,
+    concat_dimension = 0 : i64,
+    split_count = 4 : i64,
+    replica_groups = dense<[[0, 1, 2, 3]]> : tensor<1x4xi64>,
+    channel_handle = #stablehlo.channel_handle<handle = 1, type = 0>
```
Comment (GleasonK) [inline stablehlo/tests/vhlo_to_version_downgrade_invalid.mlir:94]: Is there a default channel_handle semantically speaking? I.e. is `#stablehlo.channel_handle<handle = 0, type = 0>` the same as handle omission?

Not sure if we want to support optional attrs as having a default value, or require that the attr specifically be a DefaultValuedAttr for us to support dropping a default.

File: stablehlo/dialect/VhloOps.td [hunk 3]
```diff
@@ -36,6 +36,7 @@ def VHLO_Dialect : Dialect {
              Add CollectivePermuteOp::channel_handle: https://github.com/openxla/stablehlo/pull/388.
              Add CustomCallOp::output_operand_aliases: https://github.com/openxla/stablehlo/pull/403.
              Add PartitionIdOp: https://github.com/openxla/stablehlo/pull/673.
+             Add AllToAllOp::channel_handle: TBD.
```
Comment (GleasonK) [inline stablehlo/dialect/VhloOps.td:39]: Change to https://github.com/openxla/stablehlo/pull/794 before submit

File: stablehlo/dialect/StablehloOps.td [hunk 4]
```diff
@@ -1438,9 +1438,20 @@ def StableHLO_AllToAllOp : StableHLO_Op<"all_to_all",
...
     I64Attr:$split_count,
-    I64ElementsAttr:$replica_groups
+    I64ElementsAttr:$replica_groups,
+    OptionalAttr<StableHLO_ChannelHandle>:$channel_handle
   );
   let results = (outs HLO_Tensor);
+
+  // channel_handle is only used for the SPMD partitioner, so we add a
+  // simplified builder method for convenience.
+  let builders = [
```
Comment (GleasonK) [inline stablehlo/dialect/StablehloOps.td:1448]: What does the auto-generated builder for an optional value look like? Surprised we need this for a trailing optional value

**Review by GleasonK — APPROVED** [review]

File: stablehlo/transforms/VhloToVersion.cpp (hunk 1 above)
Comment (GleasonK) [reply stablehlo/transforms/VhloToVersion.cpp:351]: I've added this to my to-do list. Just wanted to make note while reviewing.

File: stablehlo/tests/vhlo_to_version_downgrade_invalid.mlir (hunk 2 above)
Comment (burmako) [reply stablehlo/tests/vhlo_to_version_downgrade_invalid.mlir:94]: In general, I think that the situation with channel handles in MHLO/StableHLO can be improved. We have #647, #649, #666 and, more generally, #608 to sort it out. For now, I propose that we follow the established conventions in MHLO, which is what I implemented in this PR.

File: stablehlo/dialect/VhloOps.td (hunk 3 above)
Comment (burmako) [reply stablehlo/dialect/VhloOps.td:39]: Done.

File: stablehlo/dialect/StablehloOps.td (hunk 4 above)
Comment (burmako) [reply stablehlo/dialect/StablehloOps.td:1448]: Here's how it looks like (with some manual formatting):

```C++
void AllToAllOp::build(
    ::mlir::OpBuilder &odsBuilder, 
    ::mlir::OperationState &odsState, 
    ::mlir::TypeRange resultType0, 
    ::mlir::ValueRange operand, 
    /*optional*/::mlir::IntegerAttr split_dimension, 
    /*optional*/::mlir::IntegerAttr concat_dimension, 
    /*optional*/::mlir::IntegerAttr split_count, 
    ::mlir::DenseIntElementsAttr replica_groups, 
    /*optional*/::mlir::mhlo::ChannelHandleAttr channel_handle) {
  odsState.addOperands(operand);
  if (split_dimension) {
    odsState.addAttribute(getSplitDimensionAttrName(odsState.name), split_dimension);
  }
  if (concat_dimension) {
    odsState.addAttribute(getConcatDimensionAttrName(odsState.name), concat_dimension);
  }
  if (split_count) {
    odsState.addAttribute(getSplitCountAttrName(odsState.name), split_count);
  }
  odsState.addAttribute(getReplicaGroupsAttrName(odsState.name), replica_groups);
  if (channel_handle) {
    odsState.addAttribute(getChannelHandleAttrName(odsState.name), channel_handle);
  }
  odsState.addTypes(resultType0);
}
```

Comment (GleasonK) [reply stablehlo/dialect/StablehloOps.td:1448]: Ah, so optional doesn't have a default None value. LGTM in that case.

**Pull request merged**